src/devices/Nmea0183/AisManager.cs

```diff
@@ -0,0 +1,927 @@
+        {
+            lock (_lock)
+            {
+                var ship = GetOrCreateTarget<Ship>(mmsi, x => new Ship(x), lastSeenTime);
+
+                // The transceiver type is derived from the message type (a PositionReportClassA message is obviously only sent by class A equipment)
+                if (transceiverClass != AisTransceiverClass.Unknown)
+                {
+                    ship.TransceiverClass = transceiverClass;
+                }
+
+                return ship!;
+            }
+        }
+
+        private T GetOrCreateTarget<T>(uint mmsi, Func<uint, T> constructor, DateTimeOffset? lastSeenTime)
+        where T : AisTarget
+        {
+            lock (_lock)
+            {
+                AisTarget? target;
+                T? ship;
+                if (TryGetTarget(mmsi, out target) && target is Ship)
+                {
+                    ship = target as T;
+                }
+                else
+                {
+                    // Remove the existing key (this is for the rare case where the same MMSI suddenly changes type from ship to base station or similar.
+                    // That should not normally happen, but we need to be robust about it.
+                    _targets.TryRemove(mmsi, out _);
+                    ship = constructor(mmsi);
+                    _targets.TryAdd(mmsi, ship);
+                }
+
+                if (lastSeenTime.HasValue && ship != null)
+                {
+                    ship.LastSeen = lastSeenTime.Value;
+                }
+
+                return ship!;
+            }
+        }
+
+        private BaseStation GetOrCreateBaseStation(uint mmsi, AisTransceiverClass transceiverClass, DateTimeOffset? lastSeenTime)
+        {
+            return GetOrCreateTarget<BaseStation>(mmsi, x => new BaseStation(mmsi), lastSeenTime);
+        }
+
+        private SarAircraft GetOrCreateSarAircraft(uint mmsi, DateTimeOffset? lastSeenTime)
+        {
+            return GetOrCreateTarget<SarAircraft>(mmsi, x => new SarAircraft(mmsi), lastSeenTime);
+        }
+
+        /// <summary>
+        /// Gets the list of active targets
+        /// </summary>
+        /// <returns>An enumeration of all currently tracked targets</returns>
+        public IEnumerable<AisTarget> GetTargets()
+        {
+            lock (_lock)
+            {
+                return _targets.Values;
+            }
+        }
+
+        /// <summary>
+        /// Gets the list of all active targets of the given type
+        /// </summary>
+        /// <typeparam name="T">A type of target, must be a derivative of <see cref="AisTarget"/>.</typeparam>
+        /// <returns>An enumeration of all targets of that type</returns>
+        public IEnumerable<T> GetSpecificTargets<T>()
+        where T : AisTarget
+        {
+            lock (_lock)
+            {
+                return _targets.Values.Where(x => x is T).Cast<T>();
```

Comment:
You can directly use `.OfType<T>()` and remove both `.Where` and `.Cast`

Reply:
Nice! I didn't know this one.